syntax = "proto3";

service Branch {
    // Branch Services 
    rpc Deposit(DepositRequest) returns (DepositResponse);
    rpc Withdraw(WithdrawRequest) returns (WithdrawRequest);
    rpc Query(QueryRequest) returns (QueryResponse);
    rpc WithdrawPropogate(WithdrawPropogateRequest) returns (WithdrawPropogateResponse);
    rpc DepositPropogate(DepositPropogateRequest) returns (DepositPropogateRequest);

}

// API Requests
message WithdrawRequest {
    int32 id = 1;
    int32 clock = 2;
    Event event = 3;
}

message DepositRequest {
    int32 id = 1;
    int32 clock = 2;
    Event event = 3;
}

message QueryRequest {
    int32 id = 1;
    int32 clock = 2;
    Event event = 3;
}

message WithdrawPropogateRequest {
    int32 id = 1;
    int32 balance = 2;
    int32 clock = 3;
}

message DepositPropogateRequest {
    int32 id = 1;
    int32 balance = 2;
    int32 clock = 3;
}

message Event {
    int32 id = 1;
    Interface interface = 2;
    int32 money = 3;
}


//API Response
message WithdrawResponse {
    int32 id = 1;
    Interface interface = 2;
    Result result = 3;
    repeated ClockEvent clock_event = 4;
}

message DepositResponse {
    int32 id = 1;
    Interface interface = 2;
    Result result = 3;
    repeated ClockEvent clock_event = 4;
}

message QueryResponse {
    int32 id = 1;
    Interface interface = 2;
    int32 money = 3;
    Result result = 4;
}

message WithdrawPropogateResponse {
    Result result = 1;
    int32 clock = 2;
}

message DepositPropogateResponse {
    Result result = 1;
    int32 clock = 2;
}

message ClockEvent { 
    int32 id = 1;
    string name = 2;
    int32 clock = 3;
}

enum Result {
    uknown=0;
    success=1;
    failure=2;
}

enum Interface {
    undefined=0;
    deposit=1;
    withdraw=2;
    query=3;
}